graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java

```diff
@@ -0,0 +1,72 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.graphql;
+
+import java.util.List;
+
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.GraphQLError;
+import graphql.GraphqlErrorException;
+import graphql.validation.ValidationError;
+
+final class GraphqlErrorsHandlers {
+
+    static final GraphqlErrorsHandler defaultErrorsHandler =
+            (ctx, input, result, negotiatedProduceType, cause) -> {
+                if (cause != null) {
+                    // graphQL.executeAsync() returns an error in the executionResult with getErrors().
+                    // Use 500 Internal Server Error because this cause might be unexpected.
+                    final ExecutionResult error = newExecutionResult(cause);
+                    return HttpResponse.ofJson(HttpStatus.INTERNAL_SERVER_ERROR, negotiatedProduceType,
+                                               error.toSpecification());
+                }
+                final List<GraphQLError> errors = result.getErrors();
+                final HttpStatus httpStatus = graphqlErrorsToHttpStatus(errors);
+                return HttpResponse.ofJson(httpStatus, negotiatedProduceType, result.toSpecification());
+            };
+
+    private GraphqlErrorsHandlers() {}
+
+    /**
+     * Return {@link ExecutionResult} based {@link Throwable}.
+     */
+    static ExecutionResult newExecutionResult(Throwable cause) {
+        return new ExecutionResultImpl(GraphqlErrorException.newErrorException()
+                                                            .message(cause.getMessage())
+                                                            .cause(cause)
+                                                            .build());
+    }
+
+    /**
+     * Return an {@link HttpStatus} based on the specified list of {@link GraphQLError}s.
+     */
+    private static HttpStatus graphqlErrorsToHttpStatus(List<GraphQLError> errors) {
+        if (errors.isEmpty()) {
+            return HttpStatus.OK;
+        }
+        if (errors.stream().anyMatch(ValidationError.class::isInstance)) {
+            // The server SHOULD deny execution with a status code of 400 Bad Request for
+            // invalidate documentation.
+            return HttpStatus.BAD_REQUEST;
+        }
+        return HttpStatus.SERVICE_UNAVAILABLE;
```

Comment:
Q) It seems like the previous behavior was to return a 200 OK.  Is there a reason for this change?

Reply:
Because Erros in ExecutionResult is not empty.

I think the previous code returned OK in the default processing logic because the user could not handle GraphQlErros and Throwable as needed.

However, users can now handle GraphQlError and Throws. Therefore, if the user wants to implement the same behavior as before, it can be implemented by adding an ErrorHandler that returns OK for all states.

Reply:
Oops, I missed this. Thanks, @jrhee17 for pointing this out. 😄 

@Bogyie I think we have to return 200 OK because we don't want to make a breaking change.
If we change the status code, all clients that receive an error response from this service would get affected, which is a breaking change. 

We can change that only if it's a bug. But returning 200 is not a bug and some of the implementations use 200 status:
https://www.apollographql.com/docs/apollo-server/data/errors/#returning-http-status-codes
(Official spec does not define the response code by the way. 😅 )

Reply:
@jrhee17 @minwoox Oh... I don't know that. Thank you both for advising me on my mistake. 😅
Got it! I will fix as soon as possiable.